consumer/file.go

```diff
@@ -140,351 +139,118 @@ func init() {
-	spin := tsync.NewCustomSpinner(cons.source.pollingDelay)
+func (cons *File) observeFile(name string) {
+	defer cons.WorkerDone()
+	file := cons.newObservedFile(name)
+	defer file.close()
 
-	for cons.source.state != fileStateDone {
-		cons.read(buffer, sendFunction, spin.Yield, spin.Reset)
-	}
-}
+	enqueue := cons.SimpleConsumer.Enqueue
 
-func (cons *File) read(buffer *tio.BufferedReader, sendFunction func(data []byte), onEOF func(), onAfterRead func()) {
-	// Initialize the seek state if requested
-	// Try to read the remains of the file first
-	if cons.source.state == fileStateOpen {
-		if cons.source.file != nil {
-			buffer.ReadAll(cons.source.file, sendFunction)
+	if cons.hasToSetMetadata {
+		dir, file := filepath.Split(name)
+		enqueue = func(data []byte) {
+			metaData := core.Metadata{}
+			metaData.SetValue("file", []byte(file))
+			metaData.SetValue("dir", []byte(dir))
+			cons.EnqueueWithMetadata(data, metaData)
 		}
-		cons.initFile()
-		buffer.Reset(uint64(cons.seeker.offset))
 	}
 
-	// Try to open the file to read from
-	if cons.source.state == fileStateRead && cons.source.file == nil {
-		file, err := os.OpenFile(cons.source.realFileName, os.O_RDONLY, 0666)
-
-		switch {
-		case err != nil:
-			if cons.source.printFileOpenError {
-				cons.Logger.Warning("Open failed: ", err)
-				cons.source.printFileOpenError = false
-			}
-			time.Sleep(3 * time.Second)
-			return // ### continue, retry ###
-
-		default:
-			cons.source.file = file
-			cons.seeker.offset, _ = cons.source.file.Seek(cons.seeker.offset, cons.seeker.seek)
-			cons.source.printFileOpenError = true
+	if cons.offsetFilePath != "" {
+		enqueue = func(data []byte) {
+			enqueue(data)
+			file.storeOffset()
 		}
 	}
 
-	// Try to read from the file
-	if cons.source.state == fileStateRead && cons.source.file != nil {
-		err := buffer.ReadAll(cons.source.file, sendFunction)
-
-		switch {
-		case err == nil: // ok
-			onAfterRead()
-
-		case err == io.EOF:
-			if cons.source.isRotated() {
-				cons.Logger.Info("Rotation detected")
-				cons.onRoll()
-			}
-			onEOF()
-
-		case cons.source.state == fileStateRead:
-			cons.Logger.Error("Reading failed: ", err)
-			cons.source.file.Close()
-			cons.source.file = nil
-		}
+	switch cons.observeMode {
+	case observeModeWatch:
+		file.observeFSNotify(enqueue, cons.done)
+	default:
+		file.observePoll(enqueue, cons.done)
 	}
 }
 
-func (source *sourceFile) isRotated() bool {
-	if source.file.Name() != source.realFileName {
-		return true
-	}
-
-	if time.Since(source.lastStatCheck) > time.Second {
-		newStat, newStatErr := os.Stat(source.realFileName)
-		oldStat, oldStatErr := source.file.Stat()
-
-		source.lastStatCheck = time.Now()
+func (cons *File) observeFiles() {
+	defer cons.WorkerDone()
 
-		if newStatErr == nil && oldStatErr == nil && !os.SameFile(newStat, oldStat) {
-			return true
-		}
+	if !strings.ContainsAny(cons.fileName, "*?") {
+		cons.observeFile(cons.fileName) // blocking
+		return
```

Comment:
Maybe this special case could be removed to not have an additional branch to think about in the code?

Reply:
Actually not, because single file scraping is a lot simpler (the next commit will make this even clearer).